src/Servers/Kestrel/Core/src/KestrelServerOptions.cs

```diff
@@ -280,46 +306,45 @@ internal void Serialize(Utf8JsonWriter writer)
         writer.WriteEndArray();
     }
 
-    private void EnsureDefaultCert()
+    private static X509Certificate2? GetDevelopmentCertificateFromStore(ILogger<KestrelServer> logger)
     {
-        if (DefaultCertificate == null && !IsDevCertLoaded)
+        try
         {
-            IsDevCertLoaded = true; // Only try once
-            var logger = ApplicationServices!.GetRequiredService<ILogger<KestrelServer>>();
-            try
+            var cert = CertificateManager.Instance.ListCertificates(StoreName.My, StoreLocation.CurrentUser, isValid: true, requireExportable: false)
+                .FirstOrDefault();
+
+            if (cert is null)
             {
-                DefaultCertificate = CertificateManager.Instance.ListCertificates(StoreName.My, StoreLocation.CurrentUser, isValid: true, requireExportable: false)
-                    .FirstOrDefault();
-
-                if (DefaultCertificate != null)
-                {
-                    var status = CertificateManager.Instance.CheckCertificateState(DefaultCertificate, interactive: false);
-                    if (!status.Success)
-                    {
-                        // Display a warning indicating to the user that a prompt might appear and provide instructions on what to do in that
-                        // case. The underlying implementation of this check is specific to Mac OS and is handled within CheckCertificateState.
-                        // Kestrel must NEVER cause a UI prompt on a production system. We only attempt this here because Mac OS is not supported
-                        // in production.
-                        Debug.Assert(status.FailureMessage != null, "Status with a failure result must have a message.");
-                        logger.DeveloperCertificateFirstRun(status.FailureMessage);
-
-                        // Prevent binding to HTTPS if the certificate is not valid (avoid the prompt)
-                        DefaultCertificate = null;
-                    }
-                    else if (!CertificateManager.Instance.IsTrusted(DefaultCertificate))
-                    {
-                        logger.DeveloperCertificateNotTrusted();
-                    }
-                }
-                else
-                {
-                    logger.UnableToLocateDevelopmentCertificate();
-                }
+                logger.UnableToLocateDevelopmentCertificate();
+                return null;
```

Comment:
Now that this method returns a value, I thought early returns were clearer than nested ifs.  Happy to revert if people prefer the original structure.